### PR TITLE
creating test helper for scroll into view

### PR DESCRIPTION
**Pull request by abeadam**

### CODE REVIEW

**Comment by elwayman02** [comment]
There are linting failures that need to be fixed to get tests to pass.

File: addon-test-support/scoll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scoll-into-view-mock.js:4]: I don't think this is the approach we want to take. If we mock the entire scrollIntoView modifier, then nothing is actually being tested. I think a better approach would be to stub the `scrollIntoView` method on the Element prototype: https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView

We can then record the value of `this` each time the method gets called to know which element it was called on, and return a method to the user that allows them to pass in a data-test selector to assert the correct element was called.

Comment (abeadam) [reply addon-test-support/scoll-into-view-mock.js:4]: wondering if running the test without browser might run into any issue if we are mocking Element.prototype.scrollIntoView.

Comment (elwayman02) [reply addon-test-support/scoll-into-view-mock.js]: I think all our tests run with a headless browser anyway, so the APIs should still be available.

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:13]: @ygongdev @jdkahn I'm curious to get your thoughts here. Should we put the assert method on `this` or return it from the setup method?

This approach is nice as-written, because it means that invoking it (as seen on line 10 of the test below) feels like other setup functions, but it's less declarative because it becomes a magic method that you just have to know to call by reading the documentation. So, I'm a bit torn.

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:5]: We don't need to use `createSandbox`. Sinon is its own sandbox now, and cleanup happens automatically if we use `ember-sinon-qunit`.

https://sinonjs.org/releases/latest/sandbox/
https://github.com/elwayman02/ember-sinon-qunit

We already have the integration set up, so you can just call `sinon.stub` directly and don't need to use the `afterEach` hook for cleanup - it happens automatically.

https://github.com/elwayman02/ember-scroll-modifiers/blob/master/tests/test-helper.js#L13

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:7]: `sinon.replace` will work here

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:6]: instead of `stub`, I think you can do `sinon.fake`?

https://sinonjs.org/releases/latest/fakes/

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:19]: I like that you included the possibility of passing the element instance!

File: tests/integration/modifiers/scoll-into-view-mock-test.js
Comment (elwayman02) [inline tests/integration/modifiers/scoll-into-view-mock-test.js:18]: let's add a string here to describe the assertion (and for the assertion below as well)

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:22]: For this project, we prefer `let` over `const` (on this line and line 6) for variables that are not full caps CONSTANTS. 

Context: https://madhatted.com/2016/1/25/let-it-be

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:1]: We will need to mark `sinon` as a peerDependency and remember to document it as a requirement for using this test helper. We don't want to make it a true dependency, because that would require it for production builds.

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:3]: Add a newline above this function to improve the readability

File: tests/integration/modifiers/scoll-into-view-mock-test.js
Comment (elwayman02) [inline tests/integration/modifiers/scoll-into-view-mock-test.js:12]: Add a newline above this one to separate the setup functions from the tests

**Review by elwayman02 — CHANGES_REQUESTED** [review]
This is looking great! Added a few small comments just to clean things up, but I really like this approach in general.

Can you add two more tests for the following scenarios?

1) `shouldScroll` = false, correct selector passed in, test helper returns false
2) `shouldScroll` = true, incorrect selector passed in, test helper returns false

**Review by ygongdev — COMMENTED** [review]
Personally, I think `setup` are too heavy and black-boxy for simple helpers. For example, it is unclear that `scrollIntoViewCalledWith` is available when using `setupScrollIntoView` unless the user knew about it beforehand.

I think something more declarative similar to [did-intersecti-mock](https://github.com/elwayman02/ember-scroll-modifiers/blob/master/addon-test-support/did-intersect-mock.js) would be easier to understand and maintain.

```
this.scrollIntoViewMock = mockScrollIntoView(sinon);

await render(...);
this.scrollIntoViewMock.calledWith('[data-test-some-element']);
```

**Comment by elwayman02** [comment]
I like that design a lot, @ygongdev!
 
> This would also remove the need for adding `sinon` as `peerDependency` because that is the responsibility of the consumer. `restore` might also not be needed because iirc sinon's `sandbox` should restore by default.

I'm a little iffy on this. If we are asking the consumer to pass `sinon` in, that means we're essentially asking them to have it as a peerDependency without specifying it properly. If they were to pass something that wasn't sinon, it wouldn't work unless the thing they passed had the exact same API and behavior as sinon. At that point, it's better for us to just import sinon ourselves and mark it as a peerDep. The alternative here is to do the mocking and cleanup manually in the test helper without sinon, which I'm not a huge fan of either.

**Comment by abeadam** [comment]
The concern about black box is solved by the docs that I am working on. I am happy to change it though to the updated design. The reason I went with test hooks is that we need to clean up after ourselves and return the element to what it originally was.

File: addon-test-support/scroll-into-view-mock.js
Comment (jdkahn) [reply addon-test-support/scroll-into-view-mock.js:13]: @elwayman02 I'm not sure I understand the alternative you're presenting. Are you saying that we have this as a setup function that is used like this
```
hooks.beforeEach(function() {
    this.scrollIntoViewCalledWith = setupScrollIntoView();
}

hooks.afterEach(function() {
    teardownScrollIntoView();
}

test('x', async function(assert) {
    await render(...);
    assert.ok(this.scrollIntoViewCalledWith(...));
}
```

File: addon-test-support/scroll-into-view-mock.js
Comment (jdkahn) [inline addon-test-support/scroll-into-view-mock.js:25]: @abeadam @elwayman02  What are your thoughts on placing the assert inside the function here, so that 
```
assert.ok(this.scrollIntoViewCalledWith(...));
```
becomes
```
this.assertScrollIntoViewCalled(...));
```
and in the setup function we can set the assert message so the user doesn't have to.

File: tests/integration/modifiers/scoll-into-view-mock-test.js
Comment (abeadam) [reply tests/integration/modifiers/scoll-into-view-mock-test.js:18]: is this comment still valid?

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:4]: Replace all instances of `const` with `let` in this file.

**Review by elwayman02 — CHANGES_REQUESTED** [review]
This version looks great! Just one minor syntax comment, but I think this design is the right one. Can you go ahead and write up the documentation for using this test helper? Great work!

File: addon-test-support/scroll-into-view-mock.js
Comment (abeadam) [reply addon-test-support/scroll-into-view-mock.js:4]: is that so it supports older browsers?

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [reply addon-test-support/scroll-into-view-mock.js]: I would love to do that, but there's no good way to get access to the `assert` method without making the user pass it to us, at which point we should just let them do it instead.

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [reply addon-test-support/scroll-into-view-mock.js]: No, it's because these aren't constants so we shouldn't be marking them as `const`.

File: tests/integration/modifiers/scoll-into-view-mock-test.js
Comment (elwayman02) [inline tests/integration/modifiers/scoll-into-view-mock-test.js:27]: Please add descriptive text for each of these assertions to identify which behavior is being tested. That makes it easier to read test output when things fail and know exactly which assertions are breaking.

File: tests/dummy/app/templates/docs/scroll-into-view.md
Comment (elwayman02) [inline tests/dummy/app/templates/docs/scroll-into-view.md:27]:
```suggestion
* `resetMock()` - restores the native scrollIntoView function
```

File: tests/dummy/app/templates/docs/scroll-into-view.md
Comment (elwayman02) [inline tests/dummy/app/templates/docs/scroll-into-view.md:33]: Let's go ahead and use a `beforeEach`/`afterEach` pattern in this example, since that's how we expect people to use it most of the time. They can choose to do it differently, but let's show them the ideal scenario in our docs.

File: tests/dummy/app/templates/docs/scroll-into-view.md
Comment (elwayman02) [inline tests/dummy/app/templates/docs/scroll-into-view.md:36]: I don't think we need to include the extra divs here - people will be confused by them. Let's just have 1 div with the modifier and data-test selector.

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:6]: Question: Should we call the real method as well after recording the invocation element, in case the user still wants the behavior to execute for some reason? e.g. `preExistingScrollFunction.call(this, ...arguments)` or whatever.

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:9]: Typo: `preExistingScrollFunction` vs `preExistingScrollFuction`

File: addon-test-support/scroll-into-view-mock.js
Comment (elwayman02) [inline addon-test-support/scroll-into-view-mock.js:25]: See above re: typo

File: tests/dummy/app/templates/docs/scroll-into-view.md
Comment (elwayman02) [inline tests/dummy/app/templates/docs/scroll-into-view.md:46]: Let's wrap this in an `assert.ok` with a descriptor to show how to ideally use the result.

File: addon-test-support/scroll-into-view-mock.js
Comment (abeadam) [reply addon-test-support/scroll-into-view-mock.js:6]: we can, not sure if that might have undesirable side effects in some tests though assume for example that the scroll is still going when you transition from one test to the next.

Comment (elwayman02) [reply addon-test-support/scroll-into-view-mock.js:6]: I'm 50/50 either way, so I wouldn't push strongly for this; it was just a thought that occurred to me. I'm curious what @jdkahn @ygongdev think, though!

File: tests/dummy/app/templates/docs/scroll-into-view.md
Comment (elwayman02) [inline tests/dummy/app/templates/docs/scroll-into-view.md:46]: Let's do single quotes here for the descriptor. Normally the linter would catch that, but since it's an MD file the code doesn't get linted. Might as well make it clean and internally consistent, though!

**Review by elwayman02 — APPROVED** [review]
Outside of the final nit comment, I think this is ready to go from my perspective! @ygongdev @jdkahn do you want to give it a final review as well?

**Review by elwayman02 — APPROVED** [review]

File: addon-test-support/scroll-into-view-mock.js
Comment (ygongdev) [reply addon-test-support/scroll-into-view-mock.js:6]: I wouldn't call the actual function, at most I'd do something like `await triggerEvent('scroll', ...);`.
Imo, it is like how you would prefer `await click(button)` over `button.click()` in a test?

Comment (elwayman02) [reply addon-test-support/scroll-into-view-mock.js:6]: I think that would be even more confusing. The user has already triggered the method on purpose. We should either call it or not, but I don't know why we would call `triggerEvent` here.

Comment (ygongdev) [reply addon-test-support/scroll-into-view-mock.js:6]: Hmm, let me rephrase. 

We might not want to call the native `scrollIntoView` because the scroll is `non-deterministic` and it might not handle the `ember-testing` container well, which is #2.

I mention `await triggerEvent(..., 'scroll')` because it seems to mimics #2 deterministically during testing. Looks like there is also a new [`scrollTo`](https://github.com/emberjs/ember-test-helpers/blob/master/API.md#scrollto) API as well.

This also doesn't have to live internally and we can leave it up to the consumer.

Comment (elwayman02) [reply addon-test-support/scroll-into-view-mock.js:6]: Ah, I see. It might be worth adding the `scrollTo` call if we can ensure it's called in a way that would mimic scrollIntoView.

Comment (abeadam) [reply addon-test-support/scroll-into-view-mock.js:6]: I disagree with causing DOM changes in a test by default, I think it's unneeded possible source of errors. We can provide the ability for the caller to pass in a parameter or something to invoke the native method as well.

Comment (ygongdev) [reply addon-test-support/scroll-into-view-mock.js:6]: It is up to you, but if you want to provide a helper for the caller, you can just expose another function right?
```
this.mockHelperFunctions.invokeScrollIntoView(element);
```

Comment (elwayman02) [reply addon-test-support/scroll-into-view-mock.js:6]: I'm fine not adding this behavior until we have a use-case for it.

Comment (abeadam) [reply addon-test-support/scroll-into-view-mock.js:6]: @elwayman02 that would be my vote as well.

**Comment by elwayman02** [comment]
@ygongdev @jdkahn any objections to merging this as-is?

File: addon-test-support/scroll-into-view-mock.js
Comment (ygongdev) [reply addon-test-support/scroll-into-view-mock.js:6]: I'm fine with this, adding it later wouldn't be a breaking change anyways.

**Review by ygongdev — APPROVED** [review]

**Comment by elwayman02** [comment]
@abeadam sorry for the delay on merging, meant to get this done last week. Great work!

**Pull request merged**
